src/main/java/io/pravega/connectors/flink/FlinkPravegaWriter.java

```diff
@@ -55,106 +73,464 @@
+        this.txnGracePeriodMillis = txnGracePeriodMillis;
     }
 
+    // ------------------------------------------------------------------------
+
     @Override
     public void open(Configuration parameters) throws Exception {
-        this.eventSerializer = new Serializer<T>() {
-            @Override
-            public ByteBuffer serialize(T event) {
-                return ByteBuffer.wrap(serializationSchema.serialize(event));
-            }
 
-            @Override
-            public T deserialize(ByteBuffer serializedValue) {
-                throw new IllegalStateException("deserialize() called for a serializer");
-            }
-        };
-        this.writeError = new AtomicReference<>(null);
-        this.pendingWritesCount = new AtomicInteger(0);
-        ClientFactory clientFactory = ClientFactory.withScope(this.scopeName, this.controllerURI);
+        // Pravega transaction writerImpl (exactly-once) implementation can be used only when checkpoint is enabled
+        if (isCheckpointEnabled()) {
+            this.writerImpl = new TransactionWriter();
+        } else {
+            this.writerImpl = new StandardWriter();
+        }
+
+        final Serializer<T> serializer = new FlinkSerializer<>(serializationSchema);
+
+        ClientFactory clientFactory = ClientFactory.withScope(scopeName, controllerURI);
+
         this.pravegaWriter = clientFactory.createEventWriter(
-                this.streamName,
-                this.eventSerializer,
+                streamName,
+                serializer,
                 EventWriterConfig.builder().build());
-        this.executorService = Executors.newFixedThreadPool(5);
-        log.info("Initialized pravega writer for stream: {}/{} with controller URI: {}", this.scopeName,
-                 this.streamName, this.controllerURI);
+
+        log.info("Initialized pravega writerImpl for stream: {}/{} with controller URI: {}", scopeName,
+                streamName, controllerURI);
+
+        writerImpl.initialize();
+    }
+
+    @Override
+    public void invoke(T event) throws Exception {
+        writerImpl.write(event);
     }
 
     @Override
     public void close() throws Exception {
-        if (this.writerMode == PravegaWriterMode.ATLEAST_ONCE) {
-            flushAndVerify();
+        writerImpl.close();
+    }
+
+    // ------------------------------------------------------------------------
+
+    @Override
+    public List<UUID> snapshotState(long checkpointId, long checkpointTime) throws Exception {
+        return writerImpl.snapshotState(checkpointId, checkpointTime);
+    }
+
+
+    /**
+     * Restores the state, which here means the IDs of transaction for which we have
+     * to ensure that they are really committed.
+     */
+    @Override
+    public void restoreState(List<UUID> list) throws Exception {
+        // restore will be called before open. Create a local internal writer only to assist the restore functionality
+        InternalWriter<T> writerImpl;
```

Comment:
Smells bad to use a local writerImpl variable when a field of the same name exists, and to use an uninitialized writer instance.   It also seems odd that the writerImpl isn't mutated by the call to `restoreState`.

I can think of a few alternatives:
1. Move the writerImpl-related code from `open` to a shared `initializeWriter` method that would also be called from `restoreState`.   
2. use a static method on `TransactionWriter` to restore state and assert that checkpointing is enabled.   This would make most sense if `restoreState` doesn't mutate the writerImpl's state.